src/runtime/linux/auxvec.rs

```diff
@@ -0,0 +1,86 @@
+//! Reads /proc/self/auxv on Linux systems
+
+use std::prelude::v1::*;
+use std::slice;
+use std::mem;
+use std::fmt;
+
+/// Simple abstraction for the ELF Auxiliary Vector
+///
+/// the elf.h provide the layout of the single entry as auxv_t.
+/// The desugared version is a usize tag followed by a union with
+/// the same storage size.
+pub struct AuxVec {
+    raw: Vec<usize>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[allow(dead_code)]
+/// ELF Auxiliary vector entry types
+///
+/// The entry types are specified in `linux/auxvec.h`.
+pub enum AT {
+    /// CPU Hardware capabilities, it is a bitfield.
+    HWCAP = 16,
+    /// CPU Hardware capabilities, additional bitfield.
+    HWCAP2 = 26,
+}
+
+impl fmt::Debug for AuxVec {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_list()
+            .entries(self.raw.chunks(2).map(|kv| (kv[0], kv[1])))
+            .finish()
+    }
+}
+
+impl AuxVec {
+    /// Reads /proc/self/auxv into AuxVec.
+    // TODO: make it less ugly if possible
+    pub fn new() -> Result<Self, ::std::io::Error> {
+        use std::io::Read;
+        let mut file = ::std::fs::File::open("/proc/self/auxv")?;
+        let mut buf = [0usize; 64];
+        let mut raw = unsafe {
+            slice::from_raw_parts_mut(
+                buf.as_mut_ptr() as *mut u8,
+                buf.len() * mem::size_of::<usize>(),
+            )
+        };
+        let l = file.read(&mut raw)? / mem::size_of::<usize>();
+
+        mem::forget(raw);
+
+        let mut v = Vec::new();
+
+        v.extend_from_slice(&buf[..l]);
+
+        Ok(AuxVec { raw: v })
+    }
+
+    /// Returns the value for the AT key
+    ///
+    /// NOTE: Only the hardware capabilities are supported.
+    pub fn lookup(&self, key: AT) -> Option<usize> {
+        let k = key as usize;
+        for el in self.raw.chunks(2) {
+            if el[0] == k {
+                return Some(el[1]);
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
```

Comment:
I don't know if its easily possible but we should try to refactor this a bit so that we can test it by adding one or two dumps of auxv for arm cpus.

Reply:
I do not have an arm system handy right now, could somebody else do that?

Reply:
Yeah just open an issue and tag it here. Otherwise you can also try googling for dumps of auxv, maybe you get lucky :D

Reply:
Actually it won't work since the auxv entry size is system dependent so the synthetic 32bit tests would fail on 64bit platforms.

Reply:
We could use `#[cfg(target_pointer_width = "32")]` to enable the tests on 32-bit platforms or to add different tests for 32 and 64 bit platforms.

Reply:
I would not go this direction.

Reply:
IIRC 64 bit solaris still uses 32 bit auxv, but how many solaris users are out there anymore...

Reply:
BTW, since I don't run into a whole lot of people who know/care about auxval, is there anything I could have done in https://crates.io/crates/auxv to make it more likely that you would use that instead of re-rolling your own?

Reply:
I wasn't aware of that crate otherwise I'd use it assuming stdsimd is fine with external crates.

Does it survive static linking when the libc does not provide the function? From a cursory look I guess I'd add some more cfg guards and/or ask for a cfg item for the binary layout.

the stack crawler is brilliant btw :)

Reply:
Well, to be fair, who would expect someone to make a crate for such an esoteric thing... ;) The stack crawler is pretty heinous, but it _does work_, mostly... 

I'm not sure what [`__attribute__((weak))`](https://bitbucket.org/marshallpierce/rust-auxv/src/d52f19ab9d73bf379aa0eb439c842574b11f71d8/c/getauxval-wrapper.c?at=master&fileviewer=file-view-default#getauxval-wrapper.c-6) does when static linking. [This person](https://stackoverflow.com/questions/2627839/weak-linking-with-static-libraries) seems to have not had a great time with it. Is there a specific test case you had in mind? I'd be curious to see what happens in practice.

Reply:
You might try to see what happens when you build a native static library with musl (the use-case I have in mind is building VLC, we want to support more rust and the build-glue is... peculiar).

The stack crawler would be the best solution beside the "tiny" problem of being quite unsafe.

Reply:
@marshallpierce

> IIRC 64 bit solaris still uses 32 bit auxv, but how many solaris users are out there anymore...

64 bit solaris is a tier 2 platform, but even though the rest of core/std is not tested against it, external core/std components are in a position to test beyond tier 1 if they want to. 

Please fill an issue in this repo so that we don't forget this. 

---

@marshallpierce 

> BTW, since I don't run into a whole lot of people who know/care about auxval, is there anything I could have done in https://crates.io/crates/auxv to make it more likely that you would use that instead of re-rolling your own?

I honestly don't think there is anything you could have done. This is just my opinion, but in a nutshell:

- `coresimd`/`stdsimd` are intended to be part of `core`/`std` so right now AFAIK they cannot technically have any external dependencies (if these libraries weren't intended to be part of the `std` library, I think we would have used `cupid` and `auxv` directly, at least initially)
- no other language run-time that I know of aims to offer as comprehensive and safe "feature/simd" support as Rust, which means we don't have any place to draw inspiration from which makes it hard to tell a priori whether a dependency fulfills our needs. 

For example, @shepmaster also wishes that we would reuse the `cupid` crate, but back then we did not know exactly whether it would fulfill our needs (and we couldn't directly use it even if we wanted). In hindsight, now that the run-time feature detection support for `x86/x86_64` is finished, it is much easier to discuss whether we should be using `cupid` internally or not, but back then it was a big unknown. The same applies to the `auxv` crate right now. Maybe when the whole `auxv` implementation is finished we realize "actually, we could just use the `auxv` crate for this" and try to somehow position doing that somewhere in the roadmap, but right now this is a question that we don't have enough data to answer. 

This doesn't mean that there isn't anything here for the `auxv` crate though. For example, we use `cupid` as a dev dependency, which has resulted in some bug reports, bug fixes, etc. landing on `cupid` upstream. I might not be the easiest person to work with, but one of the next things we should do is add the `auxv` crate as a dev dependency and check that our results match those of the `auxv` crate.

Reply:
@lu-zero @marshallpierce I still think that `getauxval`, when available, should be our preferred way of testing features on ARM at least because of performance, availability on Android, and... because IIUC `getauxval()` does not require a `std` dependency. 

That is, if we manage to implement `getauxval()` on `coresimd`, at least on Linux, all `#[no_std]` crates would have as good run-time feature support as `x86/x86_64` on ARM/PowerPC/...

Reply:
That all makes sense. If I can find any documentation (or reproduce it in a VM) for Solaris's odd behavior, I'll file an issue. I might just be mis-remembering other Solaris oddities.

Reply:
another nice thing about `getauxval()` is that there are no syscalls; it's purely userspace. (The libc implementation is basically the same as my stack-walking code, except that they can skip the environment vars reliably and I cannot.)

Reply:
Are linux binaries guaranteed to have an `auxv` ? Or would we still need the fall back to try to read `/proc/self/auxv` and `/proc/cpuinfo` ? 

If by using your stack-walking code we could reliably read `auxv` on Linux without a libc dependency we might be able to drop the `std` component completely which would simplify many things.

Reply:
ELF binaries have an `auxv`, but I'm not sure about non-ELF things on Linux. http://phrack.org/issues/58/5.html has more info on the ELF runtime layout.

I don't think the stack walking code is suitable for a library because it couldn't necessarily guarantee that nobody has changed an environment variable, for instance. However, calling `getauxval()` (when available) is safe, and I don't see why it would require anything from the stdlib. 

`getauxval()` is available in glibc 2.16 (2012), bionic since Android 4.3 (also 2012), and musl 1.1.0 (2014). Even RHEL, champion of ancient versions, has glibc 2.17 in RHEL7. Of the target market (people using rust with simd on linux), that probably covers just about everybody, right?

Reply:
As long it is happening in the main thread and nothing plays with that memory it is quite reliable. Just it requires that nobody is changing the environment while you are walking the stack from what I recall.

Reply:
@lu-zero is there a good way to get started with rust on musl? rustup-init links against glibc, which makes it hard to get going on, say, alpine linux.

Reply:
you could unpack somewhere a Gentoo stage3 and you should have everything needed more or less.

Reply:
[this](https://github.com/korhalio/stasis) and [this](https://github.com/japaric/steed) could be the most contrived corner cases we should support decently IMHO.

Reply:
I have a branch that implements getauxval in core (https://github.com/gnzlbg/stdsimd/tree/core_auxv) and should support those use cases, the C file is only compiled and linked if the target is linux. 

@marshallpierce I've stolen the C header file from the auxv crate (the one I had before was a naive implementation without all the glibc workarounds and once I saw yours I could not unsee it) but that might introduce some licensing issues. Any chances you could also license the crate under the MIT and Apache licenses? All the code in Rust stdlib must be dual licensed under MIT/Apache so that users can choose what suits them best. Otherwise I'll switch it back to the naive weak linking implementation and we can figure out how to improve it in a license-compatible way in the future.

BTW reviews welcome.

Reply:
I've submitted a PR to make review easier #242.

Reply:
Any sufficiently good hack should be slightly bothersome, right? ;) I have no problem dual (triple) licensing. The existing license is already less restrictive than ASL2 and basically equivalent to MIT. If I put a note in the LICENSE.txt, will that appease the licensing rules? (License haggles are probably my least favorite part of open source work.)

BTW, stack crawling is broken even if nothing is concurrently modifying the environment. See the [elf_stack_broken_modified_environ](https://bitbucket.org/marshallpierce/rust-auxv/src/d52f19ab9d73bf379aa0eb439c842574b11f71d8/examples/elf_stack_broken_modified_environ.rs?at=master&fileviewer=file-view-default) example, which produces output like this for me on glibc 2.25:

```
Auxv before modifying environ:
33      140725387726848
16      3219913727
6       4096
17      100
3       94512514588736
4       56
5       10
7       140372545884160
8       0
9       94512514617760
11      1000
12      1000
13      1000
14      1000
23      0
25      140725386234953
31      140725386244032
15      140725386234969
Setting new env var, which sets environ to new array
Auxv after
```

IIRC, `setenv` changes the stack in ways we can't navigate around after the fact, but of course the libc implementation is free to hold on to the pointer straight to aux for use in `getauxval()` so that will still work.

Reply:
> If I put a note in the LICENSE.txt, will that appease the licensing rules? (License haggles are probably my least favorite part of open source work.)

IANAL but I think that adding [LICENSE_MIT](https://github.com/rust-lang-nursery/stdsimd/blob/master/LICENSE-MIT) and [LICENSE-APACHE](https://github.com/rust-lang-nursery/stdsimd/blob/master/LICENSE-APACHE) files to your repo should and mentioning that the library is triple licensed somewhere should be enough.

> BTW, stack crawling is broken even if nothing is concurrently modifying the environment. 

I don't understand how this works in general, so can't really follow why this breaks at all.

In my mental model, the arguments to the binary, environment of the shell, and auxiliary vector sections are all setup on binary initialization, and are contiguously allocated next to each other at the top of the stack. 

`argc` gives you the number of arguments, so one is able to directly jump to the environment variables section. Something like `argc` for environment variables must exist, so that one can jump to the auxiliary vector sections. 

How does adding/removing environment variables work? In my mental model adding environment variables must add them somewhere else, because otherwise if you add too many then at some point adding more would need to push all the stack down, right? Or in other words, in my mental model the section were the auxiliary vector starts cannot change.

Since my mental model doesn't fit the issue you are describing, I assume that it is probably wrong. Do you have a link that explains this?

Reply:
You're right about the stack layout: nothing will ever change that the stack has args, env, and auxv one after the other. However, once you call `setenv` with a new environment var, glibc detects that it doesn't have an entry in the stack-resident table that it can update, so it reallocs the whole env var table. Once it does that, `environ` now points somewhere far away, not into the stack, and now we can't navigate from it down the stack to auxv. However, libc conveniently keeps a pointer to the stack auxv, which is how `getauxval()` continues to work even after monkeying with env vars. 

For the gory details, you can check `stdlib/setenv.c` in a [glibc tarball](https://mirrors.syringanetworks.net/gnu/libc/).

Reply:
I added the other two licenses to `rust-auxv` so that should be good to go.

Reply:
I guess we could "just" save the auxv correct pointer on the `init()` call that is used to save the argv/argc.

At that time you are 100% sure that the layout is stable and the pointer to argv is always valid.
then you walk your way twice and you have the auxv for our purposes.

basically my idea boils down to merge an extended version of your walker in `sys::args`.

Reply:
Note that then for this to work in `core` we would still be relying on `getauxval` existing, but I guess that at that point we could just weakly-link `std` from core and use that instead.

Reply:
My idea is to avoid `getauxval` completely and just walk from argv in `imp::init()` and save a `GLOBAL_AUXV_PTR` there. then wrap the whole thing as it is done for args.

Reply:
Before we go to the extent of bypassing `getauxval()` and doing our own `unsafe` version, I think it would make sense to evaluate whether or not there is actually a problem with `getauxval()` availability in practice. Are people going to be writing Rust code that uses SIMD that gets deployed on very old Android devices? I suspect that that is likely to be a marginal use case at most.

Reply:
I assume that the `getauxval()` support is sorted out already and I looked at how hard would be avoiding it  to support the fringe cases I mentioned.

Apparently it would be within the 50-lines range.

Reply:
@lu-zero I thought that the `init` call was only available in `std`, is it also usable from `core`?

---

@marshallpierce once `getauxval` lands things will be in my opinion very good for Linux users and my time is better spent on improving things for `macos` users (macos x86 users are fine already, but `iOS` users on ARM chips are not). Having said this, if someone wants to spearhead support for libc's without `getauxval` on `coresimd` I'd completely support that (for `steed` users it might be better to just implement `getauxval` on `steed` though).

Reply:
@lu-zero if you want to pursue this a good way to get initial feedback is to fill an issue with your thoughts on the matter, we are kind of mixing many different discussions in this PR, and while many are subscribed to the issues in this repo, most of them are probably not being notified about any of this.